src/Framework/PaymentGateways/Routes/GatewayRoute.php

```diff
@@ -2,77 +2,109 @@
+        $gatewayIds = array_keys($paymentGateways);
 
-            if (!$this->isValidRequest($gatewayIds)) {
-                throw new PaymentGatewayException('This route is not valid.');
-            }
+        if ( ! $this->isValidRequest($gatewayIds, $data->gatewayId)) {
+            throw new PaymentGatewayException('This route is not valid.');
+        }
 
-            $data = GatewayRouteData::fromRequest($_GET);
+        if ( ! $this->hasValidNonce($data)) {
+            throw new PaymentGatewayException('This route does not have valid nonce.');
+        }
 
-            /** @var PaymentGateway $gateway */
-            $gateway = give($paymentGateways[$data->gatewayId]);
+        /** @var PaymentGateway $gateway */
+        $gateway = give($paymentGateways[$data->gatewayId]);
 
-            $allowedGatewayMethods = $gateway->routeMethods;
+        $allowedGatewayMethods = Gateway::isOffsitePaymentGateway($gateway) ? $gateway->routeMethods : [];
 
-            if (is_a($gateway, OffsiteGatewayInterface::class)) {
-                $allowedGatewayMethods = array_merge(
-                    $allowedGatewayMethods,
-                    OffsiteGatewayInterface::defaultRouteMethods
-                );
-            }
+        if (is_a($gateway, OffsiteGatewayInterface::class)) {
+            $allowedGatewayMethods = array_merge(
+                $allowedGatewayMethods,
+                OffsiteGatewayInterface::defaultRouteMethods
+            );
+        }
 
-            if (
-                !in_array($data->gatewayMethod, $allowedGatewayMethods, true) ||
-                !method_exists($gateway, $data->gatewayMethod)
-            ) {
-                throw new PaymentGatewayException('The gateway method does not exist.');
-            }
+        if (
+            ! in_array($data->gatewayMethod, $allowedGatewayMethods, true) ||
+            ! method_exists($gateway, $data->gatewayMethod)
+        ) {
+            throw new PaymentGatewayException('The gateway method does not exist.');
+        }
 
+        /**
+         * Gateway route can be used for:
+         * 1. Webhooks
+         * 2. Offsite payment gateway redirect
+         *
+         * Webhooks controller mostly need to return http status 200 or other, so no need to involve core logic.
+         * Payment gateway can handle it.
+         *
+         * Offsite payment gateway redirect further need redirect to success or failed or cancelled donation page.
+         * For this reason we need core to involve to handle redirect.
+         */
+        if (in_array($data->gatewayMethod, OffsiteGatewayInterface::defaultRouteMethods, true)) {
+            (new ProcessOffsitePaymentRedirectOnGatewayRoute($gateway))
+                ->handleGatewayRouteMethod($data->donationId, $data->gatewayMethod);
+        } else {
             $gateway->handleGatewayRouteMethod($data->donationId, $data->gatewayMethod);
         }
+
+        exit();
```

Comment:
@ravinderk I think we might be overcomplicating this whole thing.  We moved away from trying to compensate for every payment status command for this route handler in the latest version of the gateway API.